src/fathom.js

```diff
@@ -0,0 +1,189 @@
+        messaging.onMessage.addListener(messageListener,
+            {
+                type: "webScience.fathom.fathomData",
+                schema: {
+                    results: "object"
+                }
+            }
+        );
+        
+        // Message to send to content script if this page should be classified
+        messaging.registerSchema("webScience.fathom.isClassifiable", {
+            isClassifiable: "boolean",
+        })
+
+        // When a tab is updated, send it a message if the page should be 
+        // classified with Fathom
+        // TODO: onUpdated may not be the right event
+        browser.tabs.onUpdated.addListener((tabId, changeInfo, tab) => {
+            if ("url" in tab) {
+                // Iterate over listenerRecords, check if this url matches any
+                classifiable = false;
+                for (const listenerRecord of fathomDataListeners.values()) {
+                    if (listenerRecord.matchPatternSet.matches(tab.url)) {
+                        classifiable = true;
+                        break;
+                    }
+                }
+                // Send the message to content script
+                messaging.sendMessageToTab(tabId, {
+                    type: "webScience.fathom.isClassifiable",
+                    isClassifiable: classifiable,
+                });
+            }
+        });
+    }
+
+    // Register content script for the listener
+    const contentScript = await browser.contentScripts.register({
+        matches: matchPatterns,
+        js: [{
+            code: inline.dataUrlToString(fathomContentScript)
+        }],
+        runAt: "document_idle"
+    })
+
+    // Compile the match patterns for the listener
+    const matchPatternSet = matching.createMatchPatternSet(matchPatterns);
+
+    // Add listener to fathomDataListeners map
+    fathomDataListeners.set(listener, {
+        matchPatternSet,
+        contentScript
+    });
+}
+
+
+/**
+ * A callback function to remove a fathomDataListener.
+ * @param {fathomDataCallback} listener - The listener that is being removed.
+ * @private
+ */
+function removeListener(listener) {
+    const listenerRecord = fathomDataListeners.get(listener);
+    if (listenerRecord === undefined) {
+        return;
+    }
+    listenerRecord.contentScript.unregister();
+    fathomDataListeners.delete(listener);
+}
+
+/**
+ * A callback function to handle messages from the content script.
+ * @param {FathomDataObject} fathomDataObject - Fathom classification results.
+ * @private
+ */
+function messageListener(fathomDataObject) {
+    console.log("messageListener called");
+    for (const [listener, listenerRecord] of fathomDataListeners) {
+        listener(fathomDataObject);
+
+        // Notify listener if the url matches the listener's matchPatterns
+        if (listenerRecord.matchPatternSet.matches(fathomDataObject.url)) {
+            listener(fathomDataObject);
+        }
```

Comment:
Don't need to do this matching twice. Could associate each listener with a unique ID and include that in the message to the content script and the response.